### DISCUSSION TITLE
Query params in Nextjs pathname are being decoded

### DISCUSSION THREAD

**Discussion by Dnguye92** [vercel/next.js#22812] [Help]
We're ingesting some data from an api where we're using the url to navigate through the application. In our link component we're passing a url as `<Link href={{ pathname: url , someOtherProperty }} />` where the url looks like `/path-to-page?query=value` but when I look on the client I'm seeing that the url is being returned as `/path-to-page%3Fquery=value`. The url is fine when we aren't passing an object to `href` like `<Link href={url} />`.
`decodeURIComponent` also doesn't work when passing the url into `pathname` because it's technically correct until it's rendered to the client.

Any ideas?

**Comment by thanhtutzaw** [comment]
No Idea  I can decode url but cannot in Next/Link dynamic route . Value comes from API

**Comment by imsgx** [comment]
same problem here. any solution?

**Comment by emadbaqeri** [comment]
You are just simply seeing a proper `encoded` URL, for decoding that just simply use `decodeURLComponent()` function.

**Reply by emadbaqeri** [reply]
Actually the other way is using `useParams()` but it only works on the client-side

**Reply by emadbaqeri** [reply]
for more investigation head to [this link from MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURI)

**Comment by darthmaim** [comment]
You are setting a url with pathname + searchParams to just pathname. Try this instead:

```jsx
const href = new URL(url);
href.searchParams.set('other', 'param'); // set other params

<Link href={href}/>
```
